src/main/java/hudson/remoting/AtmostOneThreadExecutor.java

```diff
@@ -71,10 +71,12 @@ public boolean isTerminated() {
 
     public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
         synchronized (q) {
-            long start = System.currentTimeMillis();
-            long end = start+unit.toMillis(timeout);
-            while (isAlive() && System.currentTimeMillis()<end) {
-                q.wait(end-System.currentTimeMillis());
+            long now = System.currentTimeMillis();
+            long end = now + unit.toMillis(timeout);
+            while (isAlive() && now < end) {
+                q.wait(end - now);
+                // XXX this is not safe against Clock skew - but System.nanoTime() has performance implications.
```

Comment:
Hmm...
TODO should never be seen in production code IMHO - its to tag WIP as a
reminder for the contributors(s)
XXX is this is a potential area of concern so when looking for issues
around this code area pay attention to this
FIXME: it works - is hacky but was needed very quickly so needs to be
rewritten in the future - should only be present in production code for a
very short while.

I can change it - but having just one TAG is a step backwards :)  (intelliJ
supports multiple tags IIRC as does eclipse)

On 18 February 2015 at 19:05, Jesse Glick <notifications@github.com> wrote:

> In src/main/java/hudson/remoting/AtmostOneThreadExecutor.java
> <https://github.com/jenkinsci/remoting/pull/29#discussion_r24929965>:
>
> > @@ -71,10 +71,12 @@ public boolean isTerminated() {
> >
> >      public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
> >          synchronized (q) {
> > -            long start = System.currentTimeMillis();
> > -            long end = start+unit.toMillis(timeout);
> > -            while (isAlive() && System.currentTimeMillis()<end) {
> > -                q.wait(end-System.currentTimeMillis());
> > +            long now = System.currentTimeMillis();
> > +            long end = now + unit.toMillis(timeout);
> > +            while (isAlive() && now < end) {
> > +                q.wait(end - now);
> > +                // XXX this is not safe against Clock skew - but System.nanoTime() has performance implications.
>
> BTW at @stephenc <https://github.com/stephenc>’s insistence we use TODO
> for all such comments.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/remoting/pull/29/files#r24929965>.
>